etcdserver/server.go

```diff
@@ -702,7 +702,11 @@ func (s *EtcdServer) applyRequest(r pb.Request) Response {
 		switch {
 		case existsSet:
 			if exists {
-				return f(s.store.Update(r.Path, r.Val, expr))
+				if r.PrevIndex == 0 && r.PrevValue == "" {
```

Comment:
probably not. We need to make sure the code logic is readable first. 

So this is a set case actually. The cas in the case is a special one, so it should be the second case in the if statement. If we use the condition you described, then we need to get the negative of it, which complicates the issue. We can clean this whole thing up, but I do not think it worth the effort. A simple fix is enough here.